libsolidity/ast/Types.cpp

```diff
@@ -485,73 +733,110 @@ TypePointer IntegerConstantType::binaryOperatorResult(Token::Value _operator, Ty
+				value = rational(newNumerator, newDenominator);
+			}
 			else
-				value = boost::multiprecision::pow(m_value, other.m_value.convert_to<unsigned int>());
+			{
+				newNumerator = boost::multiprecision::pow(m_value.numerator(), other.m_value.numerator().convert_to<unsigned int>());
+				newDenominator = boost::multiprecision::pow(m_value.denominator(), other.m_value.numerator().convert_to<unsigned int>());
+				value = rational(newNumerator, newDenominator);
+			}
+			cout << "EXPONENT: " << to_string(value) << endl;
 			break;
+		}
 		default:
 			return TypePointer();
 		}
-		return make_shared<IntegerConstantType>(value);
+		return make_shared<ConstantNumberType>(value, scale);
 	}
 }
 
-bool IntegerConstantType::operator==(Type const& _other) const
+bool ConstantNumberType::operator==(Type const& _other) const
 {
 	if (_other.category() != category())
 		return false;
-	return m_value == dynamic_cast<IntegerConstantType const&>(_other).m_value;
+	ConstantNumberType const& other = dynamic_cast<ConstantNumberType const&>(_other);
+	return m_value == other.m_value;
 }
 
-string IntegerConstantType::toString(bool) const
+string ConstantNumberType::toString(bool) const
 {
-	return "int_const " + m_value.str();
+	if (m_value.denominator() == 1)
+		return "int_const " + m_value.numerator().str();
+	return "fixed_const " + boost::lexical_cast<string>(boost::rational_cast<double>(m_value));
 }
 
-u256 IntegerConstantType::literalValue(Literal const*) const
+u256 ConstantNumberType::literalValue(Literal const*) const
 {
 	u256 value;
 	// we ignore the literal and hope that the type was correctly determined
-	solAssert(m_value <= u256(-1), "Integer constant too large.");
-	solAssert(m_value >= -(bigint(1) << 255), "Integer constant too small.");
+	solAssert(m_value <= u256(-1), "Number constant too large.");
+	solAssert(m_value >= -(bigint(1) << 255), "Number constant too small.");
 
 	if (m_value >= 0)
-		value = u256(m_value);
+		value = u256(m_value.numerator());
 	else
-		value = s2u(s256(m_value));
+		value = s2u(s256(m_value.numerator()));
 
 	return value;
 }
 
-TypePointer IntegerConstantType::mobileType() const
+TypePointer ConstantNumberType::mobileType() const
 {
-	auto intType = integerType();
-	solAssert(!!intType, "mobileType called with invalid integer constant " + toString(false));
-	return intType;
+	if (m_value.denominator() == 1)
+	{
+		auto intType = integerType();
+		solAssert(!!intType, "mobileType called with invalid integer constant " + toString(false));
+		return intType;
+	}
+	auto fixType = fixedPointType();
+	solAssert(!!fixType, "mobileType called with invalid fixed constant " + toString(false));
+	return fixType;
 }
 
-shared_ptr<IntegerType const> IntegerConstantType::integerType() const
+shared_ptr<IntegerType const> ConstantNumberType::integerType() const
 {
-	bigint value = m_value;
+	bigint value = m_value.numerator();
```

Comment:
In fact I realized that the way this is going about it is all wrong. This needs to get rid of whatever remainder is left and truncate it.